### PR TITLE
Feature/dockerize

### PR DESCRIPTION
**Pull request by aleksandrmelnikov**

### CODE REVIEW

File: common/onepanel/base/cluster-role.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
+---
```
Comment (rushtehrani) [inline common/onepanel/base/cluster-role.yaml]: Can the same `ClusterRole` as `argo` sa be used here? Also can we just rename `argo` sa to `onepanel` if we can indeed use the same role?

File: common/onepanel/base/deploy-core-api.yaml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  template:
+    metadata:
+      name: onepanel-core
+      labels:
+        app: onepanel-core
+    spec:
+      serviceAccountName: onepanel
+      containers:
+        - name: onepanel-core
+          image: onepanel/core:1.3
```
Comment (rushtehrani) [inline common/onepanel/base/deploy-core-api.yaml]: let’s use `0.5` for this tag.

Comment (aleksandrmelnikov) [reply common/onepanel/base/deploy-core-api.yaml]: Sure thing

File: common/onepanel/base/cluster-role.yaml (hunk 1 above)
Comment (aleksandrmelnikov) [reply common/onepanel/base/cluster-role.yaml]: Will test

File: common/onepanel/base/postgres-database.yaml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: postgres
+  labels:
+    app: postgres
+  namespace: onepanel
+spec:
+  type: NodePort
```
Comment (rushtehrani) [inline common/onepanel/base/postgres-database.yaml]: I think this needs to be `ClusterIP` since we don’t expose the db outside of the cluster.

File: common/onepanel/base/postgres-database.yaml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            - name: postgres-pv-claim
+              mountPath: /var/lib/postgresql/data
+              subPath: postgres
+  volumeClaimTemplates:
+    - metadata:
+        name: postgres-pv-claim
+        labels:
+          app: postgres
+      spec:
+        storageClassName: standard
```
Comment (rushtehrani) [inline common/onepanel/base/postgres-database.yaml]: Where is this `StorageClass` defined? Is this what comes with minikube?

File: common/onepanel/base/postgres-database.yaml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    spec:
+      terminationGracePeriodSeconds: 10
+      containers:
+        - name: postgres
+          image: postgres
+          ports:
+            - containerPort: 5432
+          env:
+            - name: POSTGRES_DB
+              value: "onepanel"
```
Comment (rushtehrani) [inline common/onepanel/base/postgres-database.yaml]: This should come from `configKeyRef`

File: common/onepanel/base/postgres-database.yaml (hunk 4 above)
Comment (aleksandrmelnikov) [reply common/onepanel/base/postgres-database.yaml]: image
Correct. Think we should make this a param?? Or add our own?

Comment (rushtehrani) [reply common/onepanel/base/postgres-database.yaml]: We need to add our own, because k3s and microk8s may not call theirs `standard`, we already have a storageclass, I think that's only for GKE right now.

File: common/onepanel/base/kustomization.yaml [hunk 6]
```diff
@@ -7,9 +7,15 @@ resources:
  - add-label.yaml
  - add-network-policy.yaml
  - add-role.yaml
+ - add-service-account.yaml
+ - cluster-role.yaml
+ - cluster-role-binding.yaml
  - secrets.yaml
  - database-secret.yaml
+ - deploy-core-api.yaml
```
Comment (rushtehrani) [inline common/onepanel/base/kustomization.yaml]: we need to standardize naming on these. I see a mix of:

```
<action>-<k8s-resource-name>.yaml
<action>-<our-resource-name>.yaml
<k8s-resource-name>.yaml
```

File: common/onepanel/base/deploy-core-api.yaml (hunk 2 above)
Comment (rushtehrani) [reply common/onepanel/base/deploy-core-api.yaml]: @aleksandrmelnikov thought about this for a while, and `1.0.0-beta.1` is more appropriate. Let's switch to `1.0.0-beta.1` and not increment it while we test and only increment when we release.

File: common/onepanel/base/postgres-database.yaml (hunk 4 above)
Comment (aleksandrmelnikov) [reply common/onepanel/base/postgres-database.yaml]: Sure, sounds good. Any thoughts what I should name it? `onepanel-default`?

File: common/onepanel/base/vars.yaml [hunk 7]
```diff
@@ -19,4 +19,11 @@ database:
     default: onepanel
   driverName:
     required: true
-    default: postgres
\ No newline at end of file
+    default: postgres
+application:
+  webHost:
```
Comment (rushtehrani) [inline common/onepanel/base/vars.yaml]: let's update this section and its default like so:
```
application:
  host: <host-or-ip>
  apiPort: 32001
  uiPort: 32000
```
Do we have a way to change the key/value here based on provider, for example, if it's local, it'll be `uiPort`/`apiPort`, but if it's GKE, AKS or EKS it'll be `uiPath`, `apiPath`.

**Comment by aleksandrmelnikov** [comment]
Thank you for all the feedback.

File: common/onepanel/base/postgres-database.yaml (hunk 4 above)
Comment (rushtehrani) [reply common/onepanel/base/postgres-database.yaml]: I think `onepanel` should suffice, it's already [here](https://github.com/onepanelio/manifests/blob/master/storage/base/storage-class.yaml) with this GCP overlay [here](https://github.com/onepanelio/manifests/blob/master/storage/overlays/gcp/storage-class.yaml). But it's called `default` which is confusing.

File: common/onepanel/base/vars.yaml (hunk 7 above)
Comment (aleksandrmelnikov) [reply common/onepanel/base/vars.yaml]: The CLI command `init` updates the "params.yaml" file.
So I think we can.
We just need to know the provider ahead of time.
I'm thinking they have to pass the provider in?
`main init --provider=microk8s`?

And we output something like...
"Running init with provider microk8s..."

In manifests, we'd need to put placeholders for all the providers.

Comment (aleksandrmelnikov) [reply common/onepanel/base/vars.yaml]: Hmm, but how do we instruct GKE, AKS, to use those values?
Would `main deployment` over-write the deployment files based on provider?

Comment (rushtehrani) [reply common/onepanel/base/vars.yaml]: In case of the cloud providers, we deploy istio and virtual services with those paths.

File: common/onepanel/base/cluster-role.yaml (hunk 1 above)
Comment (aleksandrmelnikov) [reply common/onepanel/base/cluster-role.yaml]: ClusterRole from argo was used as reference, so this comment is completed.

Comment (rushtehrani) [reply common/onepanel/base/cluster-role.yaml]: @aleksandrmelnikov I actually meant use the exact same `clusterrole` and use `onepanel` as the user for argo.

Comment (rushtehrani) [reply common/onepanel/base/cluster-role.yaml]: disregard, let's keep them separate for now.

Comment (aleksandrmelnikov) [reply common/onepanel/base/cluster-role.yaml]: Ahhh okay

File: common/onepanel/base/deploy-core-api.yaml (hunk 2 above)
Comment (aleksandrmelnikov) [reply common/onepanel/base/deploy-core-api.yaml]: Made change to deployment files. Uploaded relevant tags to docker hub.

File: common/onepanel/base/postgres-database.yaml (hunk 3 above)
Comment (aleksandrmelnikov) [reply common/onepanel/base/postgres-database.yaml]: Just noting this change was made, hence why this is resolved.

File: common/onepanel/base/postgres-database.yaml (hunk 5 above)
Comment (aleksandrmelnikov) [reply common/onepanel/base/postgres-database.yaml]: Implemented.

File: common/onepanel/base/kustomization.yaml (hunk 6 above)
Comment (aleksandrmelnikov) [reply common/onepanel/base/kustomization.yaml]: Agreed. Any preference?

Comment (rushtehrani) [reply common/onepanel/base/kustomization.yaml]: what about `<k8s-kind>-<k8s-metadata-name>.yaml`?  Example: `deployment-core.yaml`, `deployment-core-ui.yaml`. 

I think otherwise we may have duplicates. Also the action is determined by k8s when we run `apply` and could be `create/update`.

Comment (aleksandrmelnikov) [reply common/onepanel/base/kustomization.yaml]: I agree with the action not being included.
For the k8s-metadata-name, does that mean we'd split up core.yaml? Since it has a service?
So we'd have service-core.yaml, and deployment-core.yaml?

Comment (rushtehrani) [reply common/onepanel/base/kustomization.yaml]: Good question, ideally yes, even though we're not doing that in some other places.

File: common/onepanel/base/postgres-database.yaml (hunk 4 above)
Comment (aleksandrmelnikov) [reply common/onepanel/base/postgres-database.yaml]: I can rename those two if you'd like? To `onepanel`?

File: common/onepanel/base/kustomization.yaml (hunk 6 above)
Comment (aleksandrmelnikov) [reply common/onepanel/base/kustomization.yaml]: Okay, I'll split those out and see how it looks.

File: common/onepanel/base/postgres-database.yaml (hunk 4 above)
Comment (rushtehrani) [reply common/onepanel/base/postgres-database.yaml]: cool, makes sense. You'd have to add a specific overlay for `minikube` and `microk8s`

Comment (aleksandrmelnikov) [reply common/onepanel/base/postgres-database.yaml]: Ohhh..okay. Good point.
I'm not sure if our CLI checks those yet.

Comment (rushtehrani) [reply common/onepanel/base/postgres-database.yaml]: Actually, just looked, for both microk8s and minikube we can use `default` storage class. For former you enable it via `microk8s.enable storage` and for latter it is already built in. cc @Vafilor 

For the cloud providers, we need to create SSD `StorageClass` and call it `onepanel` as they may already have `default` but that may be non-SSD storage and not sufficient for deep learning workflows.

File: common/onepanel/base/vars.yaml (hunk 7 above)
Comment (rushtehrani) [reply common/onepanel/base/vars.yaml]: Need to clarify something here, whatever they put in for `uiPort`, `apiPort` needs to persist in the YAML. I'm not exactly sure if it is or it's random. Also, because we need both grpc and http for api, this is the `params.yaml` structure:
```
application:
  host: <host-or-ip>
  apiHTTPPort: 32000
  apiGRPCPort: 32001
  uiPort: 32002
```

File: common/onepanel/base/deploy-core-api.yaml [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  namespace: onepanel
+spec:
+  type: NodePort
+  ports:
+    - port: 8888
+      targetPort: 8888
+      name: "http"
+    - port: 8887
+      targetPort: 8887
+      name: "rpc"
```
Comment (rushtehrani) [inline common/onepanel/base/deploy-core-api.yaml]: minor: let's call this `grpc`

File: common/onepanel/base/vars.yaml (hunk 7 above)
Comment (rushtehrani) [reply common/onepanel/base/vars.yaml]: Just confirmed that the nodePort is random, so it needs to be set based on their input above, example:
```
spec:
  type: NodePort
  ports:
    - port: 8888
      targetPort: 8888
      nodePort: < apiHTTPPort>
      name: "http"
    - port: 8887
      targetPort: 8887
      nodePort: < apiGRPCPort>
      name: "grpc"
```

File: common/onepanel/base/kustomization.yaml (hunk 6 above)
Comment (aleksandrmelnikov) [reply common/onepanel/base/kustomization.yaml]: @rushtehrani What about this case?
```apiVersion: v1
kind: Namespace
metadata:
  name: $(defaultNamespace) # replace with relevant namespace
  labels:
    istio-injection: enabled
    onepanel.io/enabled: "true"```

Comment (rushtehrani) [reply common/onepanel/base/kustomization.yaml]: @aleksandrmelnikov one idea is to just call it `namespace.yaml`, if there is another namespace in the folder, I assume it'll have a name and that will be `namespace-<meta.name>.yaml`.

File: common/onepanel/base/vars.yaml (hunk 7 above)
Comment (Vafilor) [reply common/onepanel/base/vars.yaml]: @rushtehrani @aleksandrmelnikov Just joining in here. Why don't we have a set variable for each service? E.g. 
api: http://localhost:8888 or
api: http://www.test.com/api

If we do the split variable approach, we're probably going to have to do some post-processing on the resulting files to replace the appropriate variables.

File: common/onepanel/base/kustomization.yaml (hunk 6 above)
Comment (aleksandrmelnikov) [reply common/onepanel/base/kustomization.yaml]: I don't think that'll work. What if we add another namespace with a placeholder for a name?
```name: $(auditNamespace)``` for example.

Maybe we do it like so?
```namespace-audit-namespace.yaml```?

File: common/onepanel/base/vars.yaml (hunk 7 above)
Comment (rushtehrani) [reply common/onepanel/base/vars.yaml]: I think this is actually cleaner, the only thing that we need to figure out is the gRPC port since https/http are known but gRPC doesn't have known port as far as I know.

File: common/onepanel/base/kustomization.yaml (hunk 6 above)
Comment (rushtehrani) [reply common/onepanel/base/kustomization.yaml]: yup, that'll work better.

Comment (aleksandrmelnikov) [reply common/onepanel/base/kustomization.yaml]: How about: ```<k8s-kind>-<k8s-metadata-name>-<k8s-metadata-namespace>.yaml```?
Because I ran into overlap with configmaps.

Comment (aleksandrmelnikov) [reply common/onepanel/base/kustomization.yaml]: What about this one?
```apiVersion: builtin
kind: LabelTransformer
metadata:
  name: metadataLabelTransformer
labels:
  app.kubernetes.io/name: onepanel
  app.kubernetes.io/instance: onepanel-v0.5.0
  app.kubernetes.io/managed-by: onepanel-cli
  app.kubernetes.io/component: onepanel
  app.kubernetes.io/part-of: onepanel
  app.kubernetes.io/version: v0.5.0
fieldSpecs:
- path: metadata/labels
  create: true```

File: common/onepanel/base/vars.yaml (hunk 7 above)
Comment (Vafilor) [reply common/onepanel/base/vars.yaml]: @rushtehrani for cloud providers, will it still be a path? 
Maybe we could do
api: http://....
apiGrpc: ....

File: common/onepanel/base/kustomization.yaml (hunk 6 above)
Comment (rushtehrani) [reply common/onepanel/base/kustomization.yaml]: makes sense, thinking if it makes sense to switch namespace and name: `<k8s-kind>-<k8s-metadata-namespace>-<k8s-metadata-name>.yaml`, since name is in namespace.

Comment (rushtehrani) [reply common/onepanel/base/kustomization.yaml]: I think we can leave those as they are, we'll only have one `LabelTransformer` per folder.

Comment (aleksandrmelnikov) [reply common/onepanel/base/kustomization.yaml]: Copy on LabelTransformer.

What do you mean by "name is in namespace"?
That looks like python code... :)

File: common/onepanel/base/vars.yaml (hunk 7 above)
Comment (rushtehrani) [reply common/onepanel/base/vars.yaml]: no, it won't be a path. gRPC is transported over http/2 so it's the same scheme, just a different port. I think we can just have a field for `apiGRPCPort` and default it to some port number.

Comment (Vafilor) [reply common/onepanel/base/vars.yaml]: Okay, that works.

File: common/onepanel/base/kustomization.yaml (hunk 6 above)
Comment (rushtehrani) [reply common/onepanel/base/kustomization.yaml]: I meant a resource by name of x is in namespace y

File: common/onepanel/base/vars.yaml (hunk 7 above)
Comment (Vafilor) [reply common/onepanel/base/vars.yaml]: Actually, no wait it doesn't.

If we have api: http://localhost:8888, it's difficult to replace the port with another one.
And if we go the ApiHost with different port keys, we're back to square one. 

We can make gRPC its own key, or we can go back to having several variables and having the CLI do post-processing.

Comment (rushtehrani) [reply common/onepanel/base/vars.yaml]: Not sure I follow, gRPC is for external consumption, the way these will get setup in Kubernetes is as services. In case of local k8s it'll be `service` with `nodePort` set, in case of cloud providers, it will be `service` and `virtualservice` with paths.

File: common/onepanel/base/kustomization.yaml (hunk 6 above)
Comment (aleksandrmelnikov) [reply common/onepanel/base/kustomization.yaml]: I pushed up a refactoring, but it's kind-name-namespace.
How does it look?

File: common/onepanel/base/deploy-core-api.yaml (hunk 8 above)
Comment (aleksandrmelnikov) [reply common/onepanel/base/deploy-core-api.yaml]: Made.

File: common/onepanel/base/postgres-database.yaml (hunk 4 above)
Comment (aleksandrmelnikov) [reply common/onepanel/base/postgres-database.yaml]: @rushtehrani @Vafilor Can you guys verify the microk8s storage class?
```NAME                          PROVISIONER            RECLAIMPOLICY   VOLUMEBINDINGMODE   ALLOWVOLUMEEXPANSION   AGE
microk8s-hostpath (default)   microk8s.io/hostpath   Delete          Immediate           false                  38m```
I thought it was this?

Comment (Vafilor) [reply common/onepanel/base/postgres-database.yaml]: That's what I got for microk8s. Let me verify default works.

Comment (Vafilor) [reply common/onepanel/base/postgres-database.yaml]: Default does not work out of the box.

```storageclass.storage.k8s.io "default" not found```

Comment (rushtehrani) [reply common/onepanel/base/postgres-database.yaml]: should we explicitly ask for the `defaultStorage` class? I don't think we're going to distinguish between `microk8s` and `minikube`.  They both fall under `single-node`. Another option is to actually query the storageClass when `--provider` is `single-node` and use that value. For cloud providers, we create a SSD storageclass for them.

Comment (Vafilor) [reply common/onepanel/base/postgres-database.yaml]: Query would be good.

Why not distinguish between them? It can be a simple overlay in kustomize can't it?

File: storage/overlays/eks/storage-class.yaml [hunk 9]
```diff
@@ -0,0 +1,11 @@
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: onepanel
+provisioner: kubernetes.io/aws-ebs
+parameters:
+  type: io1
```
Comment (rushtehrani) [inline storage/overlays/eks/storage-class.yaml]: I think regular `gp2` SSD volume would suffice here and we can remove `iopsPerGB`

File: storage/overlays/aks/storage-class.yaml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: onepanel
+provisioner: kubernetes.io/azure-disk
+parameters:
+  storageaccounttype: Premium_LRS
```
Comment (rushtehrani) [inline storage/overlays/aks/storage-class.yaml]: Should this be empty given this statement: "Premium VM can attach both Standard_LRS and Premium_LRS disks, while Standard VM can only attach Standard_LRS disks."?

File: cert-manager/overlays/aws/vars.yaml [hunk 11]
```diff
@@ -1,8 +1,8 @@
-cert-manager:
+certManager:
```
Comment (rushtehrani) [inline cert-manager/overlays/aws/vars.yaml]: Good catch on this name change.

**Comment by rushtehrani** [comment]
@aleksandrmelnikov for this, only the two comments about Azure/AWS storage types remain.

File: storage/overlays/aks/storage-class.yaml (hunk 10 above)
Comment (aleksandrmelnikov) [reply storage/overlays/aks/storage-class.yaml]: I don't follow, what should not be empty?

Comment (rushtehrani) [reply storage/overlays/aks/storage-class.yaml]: I'm assuming if we remove `storageaccounttype`, then we'll be able to attach this volume to any VM based on that statement. Not 100% sure though.

Comment (aleksandrmelnikov) [reply storage/overlays/aks/storage-class.yaml]: I pulled this from https://kubernetes.io/docs/concepts/storage/storage-classes/#new-azure-disk-storage-class-starting-from-v1-7-2

Comment (rushtehrani) [reply storage/overlays/aks/storage-class.yaml]: Let me rephrase, if we remove `storageaccounttype`, then this volume can attach to either VM types. If we make it a `Premium_LRS` disk, then it can only attach to Premium VMs.

Comment (aleksandrmelnikov) [reply storage/overlays/aks/storage-class.yaml]: Ahhhhh okay. Yes, will remove.

**Comment by aleksandrmelnikov** [comment]
@rushtehrani All outstanding comments should be closed
